routing/payment_lifecycle.go

```diff
@@ -31,58 +31,32 @@ type paymentLifecycle struct {
 func (p *paymentLifecycle) paymentState(payment *channeldb.MPPayment) (
 	uint32, lnwire.MilliSatoshi, lnwire.MilliSatoshi, bool, error) {
 
-	var (
-		// Keep track of the number of active shards we have
-		// in-flight...
-		activeShards uint32
-
-		// ...and the total value of all in-flight and settled shards.
-		remValue = p.totalAmount
-
-		// We calculate whether we have any fees left in our budget
-		// after subtracting the fees from active shards. Note that
-		// currently we set the fee limit to zero after a restart, so
-		// the active shards might be using more than the limit in that
-		// case.
-		feeBudget = p.feeLimit
-
-		// If the payment is already marked failed, we'll get ready to
-		// terminate the moment all shards have returned.
-		terminate = payment.FailureReason != nil
-	)
-
-	// Go through the HTLCs for this payment, determining if there
-	// are any in flight.
-	for _, a := range payment.HTLCs {
-		// This HTLC already failed, ignore.
-		if a.Failure != nil {
-			continue
-		}
+	// Fetch the total amount and fees that has already been sent in
+	// settled and still in-flight shards.
+	sentAmt, fees := payment.SentAmt()
+
+	// Sanity check we haven't sent a value larger than the payment amount.
+	if sentAmt > p.totalAmount {
+		return 0, 0, 0, false, fmt.Errorf("amount sent %v exceeds "+
+			"total amount %v", sentAmt, p.totalAmount)
+	}
 
-		// Subtract settled or in flight attempts from the
-		// remaining value.
-		amt := a.Route.ReceiverAmt()
-		if amt > remValue {
-			return 0, 0, 0, false, fmt.Errorf("amount "+
-				"sent %v exceeds remaining amount %v",
-				amt, remValue)
-		}
-		remValue -= amt
-
-		// We'll subtract the used fee from our fee budget, but allow
-		// the fees of the alrady sent shards to exceed our budget (can
-		// happen after restarts).
-		fee := a.Route.TotalFees()
-		if fee <= feeBudget {
-			feeBudget -= fee
-		} else {
-			feeBudget = 0
-		}
+	// We'll subtract the used fee from our fee budget, but allow the fees
+	// of the already sent shards to exceed our budget (can happen after
+	// restarts).
+	feeBudget := p.feeLimit
+	if fees <= feeBudget {
+		feeBudget -= fees
+	} else {
+		feeBudget = 0
+	}
 
-		// If this attempt is settled, we don't want to send
-		// more shards, we will terminate the payment.
-		if a.Settle != nil {
-			terminate = true
+	// Go through the HTLCs for this payment, and count the number of
+	// shards still in flight.
+	var activeShards uint32
```

Comment:
seems like something that could be returned from `SentAmt` since it's already executing nearly the same loop

Reply:
Added method to return inflight HTLCs, as it is also useful in other places.